Sources/Build/BuildPlan.swift

```diff
@@ -259,18 +259,64 @@ public final class ProductBuildDescription {
     }
 
     /// The objects in this product.
-    let objects: [AbsolutePath]
+    // Computed during build planning.
+    fileprivate(set) var objects: [AbsolutePath] = []
 
-    /// Any addition flags to be added. These flags are expected to be computed during build planning.
+    /// The dynamic libraries this product needs to link with.
+    // Computed during build planning.
+    fileprivate(set) var dylibs: [ProductBuildDescription] = []
+
+    /// Any additional flags to be added. These flags are expected to be computed during build planning.
     fileprivate var additionalFlags: [String] = []
 
     /// Create a build description for a product.
-    init(product: ResolvedProduct, objects: [AbsolutePath], buildParameters: BuildParameters) {
+    init(product: ResolvedProduct, buildParameters: BuildParameters) {
+        assert(product.type != .library(.none), "None type libraries should not be described.")
         self.product = product
         self.buildParameters = buildParameters
-        self.objects = objects
     }
 
+    /// Computes (lazily) and returns the target and product dependencies.
+    ///
+    /// This property returns three things:
+    ///     - staticTargets: The targets which can be linked statically in this product.
+    ///     - dylibs: The products which needs to be dynamically linked.
+    ///     - systemModules: The system module targets this product depends on.
+    public lazy var dependencies: (staticTargets: [ResolvedModule], dylibs: [ResolvedProduct], systemModules: [ResolvedModule]) = {
+        var linkLibraries: [ResolvedProduct] = []
+        var systemModules: [ResolvedModule] = []
+        // Topologically sort and iterate the targets in this product until a dependency which needs to be
+        // dynamically linked is encountered. At the end we get the list of targets we can statically link
+        // and collect the dynamic products and system modules.
+        let staticTargets = try! topologicalSort(self.product.modules) { module in
+            // Include dependencies of a target if it is contained in the product, is a library or a test.
+            guard self.product.modules.contains(module) 
+               || module.type == .library
+               || module.type == .test else { return [] }
+            // Add any system module dependency.
+            systemModules += module.dependencies.filter{ $0.type == .systemModule }
+            // Add this target's local (library) dependencies.
+            var dependencies = module.dependencies.filter{ $0.type == .library }
```

Comment:
IMHO, it would be clearer to iterate `module.dependencies` once, so that it is obvious that each entry in the list is handled in *some* fashion by this method, which should be an requirement.